datacube/index/memory/_datasets.py

```diff
@@ -91,8 +92,8 @@ def add(self, dataset: Dataset,
             persistable = self.clone(dataset, for_save=True)
             self.by_id[persistable.id] = persistable
             self.active_by_id[persistable.id] = persistable
-            if dataset.uris is not None:
-                self.locations[persistable.id] = dataset.uris.copy()
+            if dataset._uris:
```

Comment:
`self.locations` is the memory drivers internal storage repository for locations, it holds locations for all datasets - it's not exposed by the AbstractIndex API.

`dataset.uris` is the same as `dataset._uris` - it a way of  temporarily continuing to support the deprecated behaviour without constantly triggering the deprecation warning.  Because of the changes in `model/__init__.py`, `dataset.uri` doesn't tell you anything you don't get from `dataset._uris`.